### PR TITLE
Fix CallbackRegistry memory leak

### PR DESCRIPTION
**Pull request by vallsv**
## PR Summary

This PR fixes the issue #19474, and probably the issue #19345

- Plus rewrite some code for readability (like full name instead of letter, intricate loop, try-catch potentially caching many things)
- Plus coverage for the method `disconnect`

A much shorter change count be done to only fix the issue, if you like.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by vallsv** [comment]
Do you really expect a coverage of 100%? The warn from `codecov/project/tests` is most probably a dead code. So i don't know what to do with it. Safer to let it there / cleaner to remove it.

Another point is the `_is_finilizing=sys.is_finilizing`. I am pretty sure this argument is not needed, else `sys.is_finilizing` became pointless. What do you thing about removing this argument? It's the same argue as my commit 12aa238, it is not needed.

File: lib/matplotlib/cbook/__init__.py [hunk 1]
```diff
@@ -199,48 +199,61 @@ def __setstate__(self, state):
 
     def connect(self, s, func):
         """Register *func* to be called when signal *s* is generated."""
-        self._func_cid_map.setdefault(s, {})
+        signal = s
```
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py:202]: please revert this rename (or perhaps we could consider `@api._rename_parameter`ing "s" to "signal" if you really want to improve that.)

File: lib/matplotlib/cbook/__init__.py [hunk 2]
```diff
@@ -199,48 +199,61 @@ def __setstate__(self, state):
...
+        self._func_cid_map.setdefault(signal, {})
         proxy = _weak_or_strong_ref(func, self._remove_proxy)
-        if proxy in self._func_cid_map[s]:
-            return self._func_cid_map[s][proxy]
+        if proxy in self._func_cid_map[signal]:
+            return self._func_cid_map[signal][proxy]
+
```
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py:207]: leave spacing as it was

File: lib/matplotlib/cbook/__init__.py [hunk 3]
```diff
@@ -199,48 +199,61 @@ def __setstate__(self, state):
...
-        self.callbacks[s][cid] = proxy
+        self._func_cid_map[signal][proxy] = cid
+        self.callbacks.setdefault(signal, {})
+        self.callbacks[signal][cid] = proxy
         return cid
 
     # Keep a reference to sys.is_finalizing, as sys may have been cleared out
     # at that point.
-    def _remove_proxy(self, proxy, *, _is_finalizing=sys.is_finalizing):
+    def _remove_proxy(self, proxy, _is_finalizing=sys.is_finalizing):
```
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py:216]: why did you remove the keyword-onlyness?

Comment (vallsv) [reply lib/matplotlib/cbook/__init__.py:216]: Because it's a private function, and it is unused. THe callback of weakref is only one argument, no need extra 

Same for `_is_finalizing=sys.is_finalizing`, i really think it is pointless.

All this stuffs was probably useful 10 years ago? But to me, now it's noise.

File: lib/matplotlib/cbook/__init__.py (hunk 1 above)
Comment (vallsv) [reply lib/matplotlib/cbook/__init__.py:202]: Thanks, I will check your decorator. I definitely prefer renaming it. It's better for everybody, including people using this API.

File: lib/matplotlib/cbook/__init__.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/cbook/__init__.py:216]: The point is to avoid other internal places from ever accidentally passing the argument.
`_is_finalizing=sys.is_finalizing` is not pointless, this may be called during interpreter shutdown etc. as documented in the comment just above; this was added in response to an actual bug which you can search in the tracker.

**Comment by anntzer** [comment]
We prefer full coverage but allow ourselves some human judgment when pieces of code are very hard to test.

File: lib/matplotlib/cbook/__init__.py (hunk 3 above)
Comment (vallsv) [reply lib/matplotlib/cbook/__init__.py:216]: Calling private function is a bug on it's own.

And i saw the comment, about finalized. But the only meaning of this function is to be called during the finalization. If there is a bug here, it's a bug inside python interpreter itself. It's probably better to add a link to this python issue at this place of the code.

Comment (anntzer) [reply lib/matplotlib/cbook/__init__.py:216]: > Calling private function is a bug on it's own.

Not if *we* call it internally.

> it's a bug inside python interpreter itself.

No, this is documented behavior (https://docs.python.org/3/reference/datamodel.html#object.__del__, https://docs.python.org/3/glossary.html#term-interpreter-shutdown).

Please be careful before making overly broad statements.

Comment (vallsv) [reply lib/matplotlib/cbook/__init__.py:216]: Well, `_remove_proxy` most not be called outside of this class by anything. You don't want it to be triggered twice anyway.

For finalizing, i am aware of this things. I just tell that the only purpose of this function is to be called during the finalizing stage. And there is no warn about a special way to use this function in the python documentation, such warning you would set with blinking style, cause it's the only use case of this function. So, i still don't think it is needed.

But i understand you tend to be conservative.

Comment (jklymak) [reply lib/matplotlib/cbook/__init__.py:216]: You haven't begun to interact with any of the conservative members of the dev team!

However, please try and defer to @anntzer's suggestions, even if they are not your own stylistic choice.  He is the Principal Engineer on the project, and was chosen for that role because of his deep knowledge of both python and matplotlib and the respect the rest of the dev teams has for that expertise.  Its possible this could be merged despite his concerns, but it will go a lot faster if you address them :wink:.

Comment (vallsv) [reply lib/matplotlib/cbook/__init__.py:216]: Don't worry, i just argue because it's an interesting question on it's own.

I use to work with many commits in such case, cause most of them are not directly connected to the problem, and i know it. There is no problem for me to rework that stuff, but i have to find the time.

Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py:216]: I agree with @anntzer that we should keep this as a keyword-only.

File: lib/matplotlib/cbook/__init__.py [hunk 4]
```diff
@@ -199,48 +199,61 @@ def __setstate__(self, state):
...
-        for signal, proxies in list(self._func_cid_map.items()):
-            try:
-                del self.callbacks[signal][proxies[proxy]]
-            except KeyError:
-                pass
-            if len(self.callbacks[signal]) == 0:
-                del self.callbacks[signal]
-                del self._func_cid_map[signal]
+        for signal, proxy_to_cid in list(self._func_cid_map.items()):
+            cid = proxy_to_cid.pop(proxy, None)
```
Comment (tacaswell) [inline lib/matplotlib/cbook/__init__.py:221]: To check my understanding, The change on L221 and L241 are the critical changes in this PR?

Comment (vallsv) [reply lib/matplotlib/cbook/__init__.py:221]: Yes. And i have just apply the same pattern to the disconnect.

File: lib/matplotlib/cbook/__init__.py (hunk 1 above)
Comment (vallsv) [reply lib/matplotlib/cbook/__init__.py]: I have just added the rename param

File: lib/matplotlib/cbook/__init__.py (hunk 2 above)
Comment (vallsv) [reply lib/matplotlib/cbook/__init__.py]: Done

File: lib/matplotlib/cbook/__init__.py (hunk 3 above)
Comment (vallsv) [reply lib/matplotlib/cbook/__init__.py]: I have removed my commit relative to the remove of `*`

File: lib/matplotlib/cbook/__init__.py [hunk 5]
```diff
@@ -197,16 +197,17 @@ def __setstate__(self, state):
             s: {proxy: cid for cid, proxy in d.items()}
             for s, d in self.callbacks.items()}
 
-    def connect(self, s, func):
+    @_api.rename_parameter("3.3.4", "s", "signal")
```
Comment (timhoffm) [inline lib/matplotlib/cbook/__init__.py:200]:
```suggestion
    @_api.rename_parameter("3.4", "s", "signal")
```
We only do API changes in minor releases not in bugfix releases. Since 3.4 is about to be tagged really soon, this may even need to go to "3.5".

File: lib/matplotlib/cbook/__init__.py [hunk 6]
```diff
@@ -197,16 +197,17 @@ def __setstate__(self, state):
             s: {proxy: cid for cid, proxy in d.items()}
             for s, d in self.callbacks.items()}
 
-    def connect(self, s, func):
+    @_api.rename_parameter("3.4", "s", "signal")
+    def connect(self, signal, func):
         """Register *func* to be called when signal *s* is generated."""
```
Comment (tacaswell) [inline lib/matplotlib/cbook/__init__.py:202]:
```suggestion
        """Register *func* to be called when signal *signal* is generated."""
```

**Review by tacaswell — APPROVED** [review]
Modulo the change to the docsting.

Thanks @vallsv !

File: lib/matplotlib/cbook/__init__.py (hunk 6 above)
Comment (vallsv) [reply lib/matplotlib/cbook/__init__.py:202]: Obviously. Thanks a lot.

**Comment by vallsv** [comment]
Let me know if you prefer better commit history.
Come can be merged and message can be fixed.

**Comment by tacaswell** [comment]
If you don't mind squashing that would be appreciated.  We can also squash-merge on our side.

**Comment by jklymak** [comment]
@anntzer did you have time to take a quick look at this?  Though it could prob be bumped to 3.4.1 as well if folks prefer.

**Comment by tacaswell** [comment]
I would also be comfortable merging this between an rc and a final release.

**Comment by anntzer** [comment]
Can you also remove cid from pickled_cids, as suggested at the end of https://github.com/matplotlib/matplotlib/issues/19345#issuecomment-767193988?

**Comment by tacaswell** [comment]
@vallsv I took the liberty of pushing commits to this branch to address the remaining memory leak.

**Review by anntzer — APPROVED** [review]

**Comment by vallsv** [comment]
Nice. Thanks for the merge.
